File: src/fbuild/deploy/reset.py
"""
Device reset functionality for embedded platforms.

Sends platform-appropriate reset signals to connected devices via serial port.
Supports Teensy (134 baud magic), ESP32 (DTR/RTS sequence), and AVR (DTR toggle).
"""

import time

from fbuild.output import log


def reset_device(platform: str, port: str, verbose: bool) -> bool:
    """Reset an embedded device via serial port.

    Args:
        platform: Platform identifier ("teensy", "esp32", "avr", etc.)
        port: Serial port name (e.g., "COM5", "/dev/ttyUSB0")
        verbose: Whether to show verbose output

    Returns:
        True if reset succeeded, False otherwise
    """
    if platform == "teensy":
        return _reset_teensy(port, verbose)
    elif platform == "esp32":
        return _reset_esp32(port, verbose)
    elif platform == "avr":
        return _reset_avr(port, verbose)
    else:
        return _reset_generic(port, verbose)


def _reset_teensy(port: str, verbose: bool) -> bool:
    """Reset Teensy via 134 baud magic.

    Opening the serial port at 134 baud triggers the Teensy bootloader's
    soft reboot mechanism. This is the standard Teensy reset method.

    Args:
        port: Serial port name
        verbose: Whether to show verbose output

    Returns:
        True if reset succeeded, False otherwise
    """
    try:
        import serial

        if verbose:
            log(f"Teensy reset: opening {port} at 134 baud (magic reboot)")

        ser = serial.Serial(port, baudrate=134)
        time.sleep(0.1)
        ser.close()

        log(f"Reset signal sent to Teensy on {port}")
        return True

    except ImportError:
        log("Error: pyserial not installed. Run: pip install pyserial")
        return False
    except KeyboardInterrupt as ke:
        from fbuild.interrupt_utils import handle_keyboard_interrupt_properly

        handle_keyboard_interrupt_properly(ke)
        raise  # Never reached, but satisfies type checker
    except Exception as e:
        if verbose:
            log(f"Teensy 134-baud reset failed: {e}")
        log(f"Failed to reset Teensy on {port}: {e}")
        return False


def _reset_esp32(port: str, verbose: bool) -> bool:
    """Reset ESP32 via DTR/RTS sequence.

    Uses the same reset sequence as esptool to toggle the EN (reset) pin
    via the DTR/RTS lines on the USB-to-serial adapter.

    Args:
        port: Serial port name
        verbose: Whether to show verbose output

    Returns:
        True if reset succeeded, False otherwise
    """
    try:
        import serial

        if verbose:
            log(f"ESP32 reset: DTR/RTS sequence on {port}")

        ser = serial.Serial(port, baudrate=115200)

        # Reset sequence: pull EN low via RTS, then release
        ser.dtr = False
        ser.rts = True
        time.sleep(0.1)
        ser.dtr = True
        ser.rts = False
        time.sleep(0.05)
        ser.dtr = False

        ser.close()

        log(f"Reset signal sent to ESP32 on {port}")
        return True

    except ImportError:
        log("Error: pyserial not installed. Run: pip install pyserial")
        return False
    except KeyboardInterrupt as ke:
        from fbuild.interrupt_utils import handle_keyboard_interrupt_properly

        handle_keyboard_interrupt_properly(ke)
        raise  # Never reached, but satisfies type checker
    except Exception as e:
        log(f"Failed to reset ESP32 on {port}: {e}")
        return False


def _reset_avr(port: str, verbose: bool) -> bool:
    """Reset AVR/Arduino via DTR toggle.

    Standard Arduino reset: toggling DTR pulses the reset pin through
    the capacitor on the board.

    Args:
        port: Serial port name
        verbose: Whether to show verbose output

    Returns:
        True if reset succeeded, False otherwise
    """
    try:
        import serial

        if verbose:
            log(f"AVR reset: DTR toggle on {port}")

        ser = serial.Serial(port, baudrate=115200)
        ser.dtr = False
        time.sleep(0.1)
        ser.dtr = True
        time.sleep(0.1)
        ser.close()

        log(f"Reset signal sent to AVR on {port}")
        return True

    except ImportError:
        log("Error: pyserial not installed. Run: pip install pyserial")
        return False
    except KeyboardInterrupt as ke:
        from fbuild.interrupt_utils import handle_keyboard_interrupt_properly

        handle_keyboard_interrupt_properly(ke)
        raise  # Never reached, but satisfies type checker
    except Exception as e:
        log(f"Failed to reset AVR on {port}: {e}")
        return False


def _reset_generic(port: str, verbose: bool) -> bool:
    """Reset device via generic DTR toggle.

    Fallback reset method that works for most devices with DTR-based reset.

    Args:
        port: Serial port name
        verbose: Whether to show verbose output

    Returns:
        True if reset succeeded, False otherwise
    """
    try:
        import serial

        if verbose:
            log(f"Generic reset: DTR toggle on {port}")

        ser = serial.Serial(port, baudrate=115200)
        ser.dtr = False
        time.sleep(0.1)
        ser.dtr = True
        time.sleep(0.1)
        ser.close()

        log(f"Reset signal sent to device on {port}")
        return True

    except ImportError:
        log("Error: pyserial not installed. Run: pip install pyserial")
        return False
    except KeyboardInterrupt as ke:
        from fbuild.interrupt_utils import handle_keyboard_interrupt_properly

        handle_keyboard_interrupt_properly(ke)
        raise  # Never reached, but satisfies type checker
    except Exception as e:
        log(f"Failed to reset device on {port}: {e}")
        return False

File: crates/fbuild-deploy/src/reset.rs
//! Device reset via serial port DTR/RTS toggling.
//!
//! Platform-specific reset sequences:
//! - Teensy: Open at 134 baud (magic reboot trigger)
//! - ESP32: DTR/RTS sequence (same as esptool)
//! - AVR/Generic: DTR toggle

use fbuild_core::Result;
use std::time::Duration;

/// Reset a device on the given serial port using a platform-appropriate sequence.
///
/// `platform` should be one of: "teensy", "esp32", "avr", "generic".
pub fn reset_device(platform: &str, port: &str, verbose: bool) -> Result<bool> {
    if verbose {
        tracing::info!("resetting {} device on {}", platform, port);
    }

    let result = match platform {
        "teensy" => reset_teensy(port, verbose),
        "esp32" => reset_esp32(port, verbose),
        "avr" => reset_avr(port, verbose),
        _ => reset_generic(port, verbose),
    };

    match &result {
        Ok(true) => {
            if verbose {
                tracing::info!("device reset successful");
            }
        }
        Ok(false) => tracing::warn!("device reset reported failure"),
        Err(e) => tracing::error!("device reset error: {}", e),
    }

    result
}

/// Detect platform string from a board identifier (from platformio.ini).
pub fn detect_platform_for_reset(board: &str) -> &'static str {
    let board_lower = board.to_lowercase();
    if board_lower.starts_with("teensy") {
        "teensy"
    } else if board_lower.starts_with("esp32") {
        "esp32"
    } else if board_lower == "uno"
        || board_lower == "nano"
        || board_lower == "mega"
        || board_lower.starts_with("atmega")
        || board_lower.starts_with("arduino")
    {
        "avr"
    } else {
        "generic"
    }
}

/// Teensy reset: open serial at 134 baud (magic reboot baud rate).
fn reset_teensy(port: &str, verbose: bool) -> Result<bool> {
    if verbose {
        tracing::info!("teensy reset: opening {} at 134 baud", port);
    }

    let _serial = serialport::new(port, 134)
        .timeout(Duration::from_secs(2))
        .open()
        .map_err(|e| {
            fbuild_core::FbuildError::SerialError(format!("failed to open {}: {}", port, e))
        })?;

    // Hold open briefly — the Teensy bootloader triggers reboot on 134 baud connection
    std::thread::sleep(Duration::from_millis(100));
    // Port drops here, closing the connection

    Ok(true)
}

/// ESP32 reset: DTR/RTS sequence (same as esptool boot sequence).
fn reset_esp32(port: &str, verbose: bool) -> Result<bool> {
    if verbose {
        tracing::info!("esp32 reset: DTR/RTS sequence on {}", port);
    }

    let mut serial = serialport::new(port, 115_200)
        .timeout(Duration::from_secs(2))
        .open()
        .map_err(|e| {
            fbuild_core::FbuildError::SerialError(format!("failed to open {}: {}", port, e))
        })?;

    // EN pin low (reset)
    serial
        .write_data_terminal_ready(false)
        .map_err(map_serial)?;
    serial.write_request_to_send(true).map_err(map_serial)?;
    std::thread::sleep(Duration::from_millis(100));

    // EN pin high (release reset)
    serial.write_data_terminal_ready(true).map_err(map_serial)?;
    serial.write_request_to_send(false).map_err(map_serial)?;
    std::thread::sleep(Duration::from_millis(50));

    // Final state
    serial
        .write_data_terminal_ready(false)
        .map_err(map_serial)?;

    Ok(true)
}

/// AVR reset: DTR toggle at 115200 baud.
fn reset_avr(port: &str, verbose: bool) -> Result<bool> {
    if verbose {
        tracing::info!("avr reset: DTR toggle on {}", port);
    }

    let mut serial = serialport::new(port, 115_200)
        .timeout(Duration::from_secs(2))
        .open()
        .map_err(|e| {
            fbuild_core::FbuildError::SerialError(format!("failed to open {}: {}", port, e))
        })?;

    serial
        .write_data_terminal_ready(false)
        .map_err(map_serial)?;
    std::thread::sleep(Duration::from_millis(100));
    serial.write_data_terminal_ready(true).map_err(map_serial)?;
    std::thread::sleep(Duration::from_millis(100));

    Ok(true)
}

/// Generic reset: same as AVR (DTR toggle).
fn reset_generic(port: &str, verbose: bool) -> Result<bool> {
    if verbose {
        tracing::info!("generic reset: DTR toggle on {}", port);
    }
    reset_avr(port, verbose)
}

fn map_serial(e: serialport::Error) -> fbuild_core::FbuildError {
    fbuild_core::FbuildError::SerialError(format!("serial control error: {}", e))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_detect_platform_for_reset() {
        assert_eq!(detect_platform_for_reset("teensy40"), "teensy");
        assert_eq!(detect_platform_for_reset("teensy41"), "teensy");
        assert_eq!(detect_platform_for_reset("esp32dev"), "esp32");
        assert_eq!(detect_platform_for_reset("esp32-c3"), "esp32");
        assert_eq!(detect_platform_for_reset("uno"), "avr");
        assert_eq!(detect_platform_for_reset("nano"), "avr");
        assert_eq!(detect_platform_for_reset("mega"), "avr");
        assert_eq!(detect_platform_for_reset("atmega328p"), "avr");
        assert_eq!(detect_platform_for_reset("unknown_board"), "generic");
    }
}
